enhancements/network/ingress-node/ingress-node-firewall-with-ebpf.md

```diff
@@ -0,0 +1,480 @@
+- in order to fail safe and avoid cluster lockup, various pinholes will be open in the firewall to allow basic services.
+
+### Workflow Description
+
+- the following diagram shows how node ingress firewall configuration is handled
+
+![Ingress Firewall configuration processing](./ingress-nodefw-config-flow.png)
+
+- the following diagram shows eBPF how ingress firewall kernel hook will process packets and apply firewall rules
+
+![eBPF workflow](./ebpf-packet-flow.png)
+
+- the following table shows sample of fail safe rules to make sure users won't block basic cluster services by configuring invalid firewall rules
+
+| Source CIDR | Protocol | Port | Purpose |
+|-------------|----------|------|---------|
+|  0.0.0.0/0  |   TCP    |  22  |  SSH    |
+|  0.0.0.0/0  |   UDP    |  68  | DHCP    |
+|  0.0.0.0/0  |   TCP    | 6443 | Kubernetes API server access |
+|  0.0.0.0/0  |   TCP    | 2380 & 2379 | Etcd access |
+
+### Implementation Details/Notes/Constraints
+
+New downstream OLM operator (openshift/ingress-nodefw-operator) will be designed to do the following:-
+- APIs for the new objecs.
+- controller for each of the new resources to handle add/update/delete
+- verification webhook to validate the configuration.  
+- library to construct and load eBPF ingess maps tables based on the provisioned rules.
+- library for statistics collection and exporting to prometheus.
+
+New downstream repo for the kernel hook implementation (openshift/ingress-nodefw-kernel)
+
+- sample `NodeEndPoint` object
+
+```yaml
+kind: NodeEndpoint
+metadata:
+  name: worker-all-interfaces
+  labels:
+    role: k8s-worker
+spec:
+  interfaces: 
+
+---
+
+kind: NodeEndpoint
+metadata:
+  name: worker-eth0
+  labels:
+    role: k8s-worker
+spec:
+  interfaces: 
+  - "eth0"
+
+---
+```
+
+- sample `IngressNodeFirewall` object
+
+```yaml
+kind: IngressNodeFirewall
+metadata:
+  name: worker-nodes-firewall-on-eth0
+  annotations:
+    node-endpoint: worker-eth0
+spec:
+  nodeSelector: has(do-node-ingress-firewall)
+  ingress:
+  - fromCIDRS:
+       - 1.1.1.1/24
+       - 100:1::1/64
+    rules:
+    - order: 10  
+      protocol: tcp
+      ports: [100]
+      action: allow
+    - order: 20
+      protocol: udp
+      ports: [8000, 8080, 9000]
+      action: allow
+  - fromCIDRS:
+       - 3.3.3.3/24
+    rules:
+    - order: 10
+      protocol: icmp
+      type: 3 # Destination unreachable
+      code: 1 # Host unreachable
+      action: allow
+  - fromCIDRS:
+       - 10:10::1/64
+    rules:
+    - order: 10
+      protocol: icmpv6
+      type: 128 # Ping request
+      action: deny
+  - fromCIDRS:
+       - 0.0.0.0/0
+       - 0::0/0
+    rules:
+    - order: 10
+      protocol: tcp
+      action: allow
+
+---
+
+kind: IngressNodeFirewall
+metadata:
+  name: worker-nodes-firewall
+  annotations:
+    node-endpoint: worker-all-interfaces
+spec:
+  nodeSelector: has(do-node-ingress-firewall)
+  ingress:
+  - fromCIDRS:
```

Comment:
I would find it more logical to have the fromCIDRs as part of the rule, and make rule a top level object underneath ingress:
~~~
ingress:
  - priority: 30
    protocol: udp
    ports:
      - 8000 - 9000
      - 9999
    fromCIDRS:
      - 1.1.1.0/24
         192.0.2.0/24
    action: deny
~~~

Reply:
I arrange that way because for the same CIDR u can have many rules specially the dp is key is the cidr to get to rules list